manager/manager.go

```diff
@@ -483,6 +515,105 @@ func (m *Manager) Stop(ctx context.Context) {
 	// mutex is released and Run can return now
 }
 
+func (m *Manager) joinAndStart(ctx context.Context) error {
+	return m.config.SecurityConfig.KeyWriter().UpdateHeaders(func(headers map[string]string, kek []byte) error {
+		if err := m.keyRotator.LoadFromHeaders(headers, kek); err != nil {
+			return err
+		}
+
+		// if there is no DEK, create a new one
+		dekStr, foundDEK := headers[defaultRaftDEKKey]
+		if !foundDEK {
+			newDek := make([]byte, 32)
+			_, err := rand.Read(newDek)
+			if err != nil {
+				return err
+			}
+			if headers[defaultRaftDEKKey], err = encodePEMHeaderValue(newDek, kek); err != nil {
+				return err
+			}
+			return m.raftNode.JoinAndStart(ctx, newDek)
+		}
+
+		if pendingKey := m.keyRotator.GetNewKey(); pendingKey != nil {
+			// if there is a pending header, try decrypting with that first - if that doesn't work, fall back
+			// on the normal dek but finish rotating
+			err := m.raftNode.JoinAndStart(ctx, pendingKey)
+			switch err.(type) {
+			case encryption.ErrCannotDecode:
+				// if the dek hasn't been used to snapshot, fall back to the original dek - a rotation
+				// still needs to happen
+				break
+			case nil:
+				// we've successfully joined - that means the previous snapshot had succeeded - finish
+				// the rotation
+				return m.keyRotator.FinishKeyRotation(pendingKey)
+			default:
+				// we failed to join but not due to a decoding error
+				return err
+			}
+		}
+
+		// there is a regular header - use that to decrypt and start
+		dek, err := decodePEMHeaderValue(dekStr, kek)
+		if err != nil {
+			return err
+		}
+		return m.raftNode.JoinAndStart(ctx, dek)
+	})
+}
+
+func (m *Manager) watchForKEKChanges(ctx context.Context) {
+	clusterWatch, clusterWatchCancel := state.Watch(m.raftNode.MemoryStore().WatchQueue(),
+		state.EventUpdateCluster{},
+	)
+	securityConfig := m.config.SecurityConfig
+	// we are our own peer from which we get certs - try to connect over the local socket
+	r := remotes.NewRemotes(api.Peer{Addr: m.Addr().String(), NodeID: securityConfig.ClientTLSCreds.NodeID()})
+	defer clusterWatchCancel()
+	for {
+		select {
+		case event := <-clusterWatch:
+			clusterEvent := event.(state.EventUpdateCluster)
+			kek := clusterEvent.Cluster.Spec.EncryptionConfig.ManagerUnlockKey
+			switch {
+			case bytes.Equal(kek, m.keyRotator.GetCurrentKEK()):
+				// if the KEK hasn't changed, do nothing
+				break
+			case m.keyRotator.GetCurrentKEK() == nil:
+				// it was previously unencrypted, and now needs to be encrypted - trigger
+				// a full rotation of the TLS keys too
+				if err := ca.RenewTLSConfigNow(ctx, securityConfig, r, nil); err != nil {
+					// try again
+					clusterWatch <- clusterEvent
+					break
+				}
+				// Now that we have a new TLS certificate, let's load it from disk and
+				// update our key rotator
+				headers, _, err := securityConfig.KeyReader().ReadHeaders()
+				if err != nil {
+					// try again
+					clusterWatch <- clusterEvent
```

Comment:
I believe the channels returned by `state.Watch` are unbuffered, so I don't see how this wouldn't deadlock.